File: engine/room.go
package engine

import (
	"math"
	"path/filepath"
	"strings"
)

// ---------------------------------------------------------------------------
// How much one search can read
//
// A search sends the transcript of its window to the model in one request,
// and one request can only hold so much. That is a number of tokens, set by
// the model and by what the answer needs, not a number of minutes. So the
// room is worked out here, from the model that is going to be asked, and
// the app turns it into the longest window the range picker lets you draw.
// A window that fits is sent as it is, in one request, with nothing split
// behind anybody's back. A window that does not fit is not drawn.
//
// Characters are the unit, because they can be counted before anything is
// sent and tokens cannot. Each way of planning has its own rate between
// the two, and the rate is the one the rest of the engine already goes by.
// ---------------------------------------------------------------------------

// Room is how much numbered transcript one search can send.
type Room struct {
	// Chars is the most characters of numbered transcript, as the model
	// reads it, one request can carry beside everything else it sends.
	Chars int `json:"chars"`
	// Tokens is the same in tokens, which is what the limit is.
	Tokens int `json:"tokens"`
	// By is what sets it: "context" for what the model can hold at once,
	// "memory" for what this machine can hold it with, "budget" for what
	// one search may cost.
	By string `json:"by"`
}

// localCharsPerToken is how many characters of German make a token for the
// local models. It is the rate the model's context is sized by as well, so
// what the room lets through is what the context is made for.
const localCharsPerToken = 2.5

// contextCeiling is the largest context llama-server is started with.
const contextCeiling = 262144

// unknownContext is what a local model this engine does not know is taken
// to hold. It is below the smallest any of the models it does know holds,
// so a window drawn for it is one every model can read.
const unknownContext = 32768

// localAnswerRoom is what a local search keeps free in the context for the
// answer, the thinking before it included. The thinking is written inside
// the answer's allowance, so it takes nothing on top.
func localAnswerRoom(maxTokens int) int {
	return min(maxTokens, 16384) + 2048
}

// LocalContext is the most tokens a local model can hold at once: what its
// file says it was made to hold, and never more than the context the engine
// starts it with. A model file this engine does not know is taken at less
// than the smallest of the ones it does.
func LocalContext(model string) int {
	name := filepath.Base(model)
	for _, m := range LanguageModels() {
		if strings.EqualFold(m.Name, name) && m.Context > 0 {
			return min(m.Context, contextCeiling)
		}
	}
	return unknownContext
}

// machineMemory is how much memory this machine has. It is a variable so
// a test can be any machine.
var machineMemory = MachineMemory

// MemoryContext is the largest context a search with this model can have
// on a machine with this much memory, by the rule the app offers models
// by, see FitsIn. A context that leaves the headroom free is always fine.
// A model that only fits tight was offered for a search of judgedContext,
// so that much it may have, and no more. A machine that does not say how
// much it has gets judgedContext too: nothing is promised, and nothing
// larger than what every model was judged at is asked of it.
func MemoryContext(m LanguageModel, total int64) int {
	if total <= 0 {
		return judgedContext
	}
	well := largestContext(func(ctx int) bool { return m.NeedsAt(ctx)+memoryHeadroom <= total })
	tight := min(largestContext(func(ctx int) bool { return m.NeedsAt(ctx) <= total }), judgedContext)
	return max(well, tight)
}

// largestContext is the largest context up to contextCeiling that fits,
// or zero when none does. What a context takes only grows with it.
func largestContext(fits func(int) bool) int {
	if !fits(0) {
		return 0
	}
	lo, hi := 0, contextCeiling
	for lo < hi {
		mid := (lo + hi + 1) / 2
		if fits(mid) {
			lo = mid
		} else {
			hi = mid - 1
		}
	}
	return lo
}

// localHeld is the most tokens a search with this local model can have on
// this machine, and what sets it: "context" for what the model holds,
// "memory" for what the machine does. A model the engine does not know the
// shape of is held by its context alone, which for such a model is small.
func localHeld(model string) (int, string) {
	held := LocalContext(model)
	m, known := LanguageModelByName(filepath.Base(model))
	if !known {
		return held, "context"
	}
	if fits := MemoryContext(m, machineMemory()); fits < held {
		return fits, "memory"
	}
	return held, "context"
}

// localContextFor is the context a local model is started with for a
// prompt this long: what contextFor asks for, never more than the model
// holds or the machine has memory for. The room keeps every prompt inside
// both, so this never cuts a prompt short. It keeps a model from being
// started with a context it was not made for, and the rounding up in
// contextFor from asking the machine for memory the room never checked.
func localContextFor(model string, promptChars, maxTokens int) int {
	held, _ := localHeld(model)
	// Never nothing: llama-server takes a context of 0 to mean the whole
	// of what the model holds, which is the very memory this keeps it
	// from asking for. A model that has no room at all is refused before
	// it is started, and this only catches it being loaded ahead of that.
	return max(min(contextFor(promptChars, maxTokens), held), 4096)
}

// SearchRoom is the room a search made with these options has. It is what
// the app asks before a window is drawn, and the same sum BuildPlan checks
// a window against before anything is sent.
func SearchRoom(opts Options, logDir string) Room {
	plan := PlanOptions{Count: opts.Count, MinLen: opts.Min, MaxLen: opts.Max,
		Context: opts.Context, Model: opts.Model, MaxTokens: opts.MaxTokens,
		Budget: opts.Budget, LogDir: logDir}
	if opts.Planner != "api" {
		model := opts.LLMModel
		if model == "" && opts.LLMURL == "" {
			model, _ = DefaultLocalModel()
		}
		plan.Local = &LocalModel{Model: model, URL: opts.LLMURL}
	}
	return planRoom(plan)
}

// planRoom works the room out. Whatever the request sends besides the
// transcript, the instructions and the ask, comes off the top.
func planRoom(opts PlanOptions) Room {
	recipe := opts.recipe()
	around := runeLen(recipe.System) + runeLen(recipe.Request(nil, nil, opts))
	if opts.Local != nil {
		// A server somebody started themselves has the context and the
		// memory they gave it, which the engine cannot know.
		held, by := contextCeiling, "context"
		if opts.Local.URL == "" || opts.Local.Model != "" {
			held, by = localHeld(opts.Local.Model)
		}
		tokens := max(held-localAnswerRoom(opts.MaxTokens), 0)
		return Room{Chars: charsFor(tokens, localCharsPerToken, around), Tokens: tokens, By: by}
	}

	// The API. The answer is given room for the retry as well: a model that
	// thinks its way through the whole allowance is asked again with twice
	// as much, and that second ask has to fit in the same context.
	facts := FactsFor(opts.Model)
	answer := max(min(opts.MaxTokens*2, facts.MaxOutput), opts.MaxTokens)
	room := Room{Tokens: max(facts.Context-answer, 0), By: "context"}
	perToken, output := apiRate(opts.LogDir)
	if facts.Priced {
		// The same sum BuildPlan uses for what a search is likely to cost:
		// the answer it usually writes, and whatever is left of the budget
		// for reading.
		left := opts.Budget - output/1e6*facts.PriceOut
		affordable := max(int(math.Floor(left/facts.PriceIn*1e6+1e-6)), 0)
		if affordable < room.Tokens {
			room.Tokens, room.By = affordable, "budget"
		}
	}
	room.Chars = charsFor(room.Tokens, perToken, around)
	return room
}

// apiRate is how many characters make a token through the API and how long
// an answer usually is, measured on this episode when it has been searched
// through the API before and the engine's own starting points when not.
func apiRate(logDir string) (float64, float64) {
	const sample = 1_000_000
	tokens, output, _ := estimateUsage(logDir, sample, "plan")
	return sample / float64(max(tokens, 1)), float64(output)
}

func charsFor(tokens int, perToken float64, around int) int {
	return max(int(float64(tokens)*perToken)-around, 0)
}

// Holds says whether a window has room for count clips of at least least
// seconds each, one after another. Clips can overlap, but a search that
// needs them to in order to find as many as it was asked for is asking for
// more than the window has in it. A hundredth of a second is let go, so a
// window drawn on a round step is not refused over rounding.
func Holds(w Window, count int, least float64) bool {
	return float64(count)*least <= w.End-w.Start+0.01
}

// LineWeight is one line of the transcript as the room counts it: when it is
// said, and how many characters it takes in the numbered transcript the
// model reads, its line break included.
type LineWeight struct {
	Start float64 `json:"start"`
	End   float64 `json:"end"`
	Chars int     `json:"chars"`
}

// WeighLines is the transcript cut into the lines a search would send, each
// with what it weighs. The app adds them up over a window to know whether
// the window fits the room, without asking the engine at every step of a
// drag. The numbers in front of a line depend on where the window starts,
// so a line is counted with the widest number any window of this episode
// would give it, which only ever errs towards a window that fits.
func WeighLines(t *Transcript) []LineWeight {
	lines := BuildLines(t.Words, t.Levels(), nil)
	out := make([]LineWeight, 0, len(lines))
	widest := len(itoa(len(lines)))
	for i, row := range strings.Split(AnnotateLines(lines), "\n") {
		if i >= len(lines) {
			break
		}
		// The number this line has here, and the widest it could have.
		pad := widest - len(itoa(lines[i].Index))
		out = append(out, LineWeight{Start: lines[i].Start(), End: lines[i].End(),
			Chars: runeLen(row) + 1 + max(pad, 0)})
	}
	return out
}

// SpokenChars is how many characters of numbered transcript a second of an
// episode makes, before its own transcript can say: a half hour of German
// came to about 40,000, and this leaves room to spare.
const SpokenChars = 25.0

// RateOf is how many characters a second of this episode makes, from the
// lines it has so far, with 15 percent on top for a stretch busier than the
// average. Until there are ten minutes of them the starting point stands,
// and it never goes below it: the rate is used for what has not been heard
// yet, and guessing that lighter than it turns out is what would let a
// window through that the engine then refuses.
func RateOf(lines []LineWeight) float64 {
	if len(lines) == 0 {
		return SpokenChars
	}
	span := lines[len(lines)-1].End - lines[0].Start
	if span < 600 {
		return SpokenChars
	}
	chars := 0
	for _, l := range lines {
		chars += l.Chars
	}
	return max(float64(chars)/span*1.15, SpokenChars)
}

File: frontend/src/lib/room.ts
// How much of an episode one search can read, turned into seconds.
//
// A search sends its window to the model in one request, and one request
// holds only so much: a number of characters of transcript, worked out by
// the engine from the model that is going to read it (see engine/room.go).
// The engine also weighs every line the transcript has so far. Here those
// weights are added up over a window, so the range picker knows while it
// is dragged how far a window may reach, without asking the Go side at
// every step of a hand. The engine checks the same sum again before it
// sends anything.

export interface LineWeight {
  start: number;
  end: number;
  chars: number;
}

export interface RoomView {
  // The most characters of transcript one request can carry.
  chars: number;
  // What sets it: "context" for what the model holds at once, "memory" for
  // what this machine can hold it with, "budget" for what one search may
  // cost.
  by: string;
  lines: LineWeight[];
  // How far the transcript reaches, silence at its end included.
  heard: number;
  // Characters a second, for the part of the episode not heard yet.
  rate: number;
}

// Bisection runs this many times. Forty halvings of a four hour episode
// are far below a millisecond.
const halvings = 40;

export class Reach {
  private readonly lines: LineWeight[];
  // sums[i] is the weight of the lines before line i.
  private readonly sums: number[];
  // Where the transcript ends. Past it, a second weighs the rate.
  private readonly heard: number;

  constructor(
    readonly room: RoomView | null,
    readonly duration: number,
  ) {
    this.lines = room?.lines ?? [];
    this.sums = [0];
    for (const line of this.lines) this.sums.push(this.sums[this.sums.length - 1] + line.chars);
    const last = this.lines.length ? this.lines[this.lines.length - 1].end : 0;
    this.heard = Math.max(room?.heard ?? 0, last);
  }

  // Whether there is a room at all. Until the engine has answered there is
  // nothing to hold a window to.
  get known(): boolean {
    return !!this.room && this.room.chars > 0;
  }

  // What a window from..to sends. A line the window only touches is
  // counted whole, which only ever errs towards a window that fits.
  weight(from: number, to: number): number {
    if (to <= from) return 0;
    // The first line that ends after from, and the first that starts at
    // or after to.
    const first = this.search((i) => this.lines[i].end > from);
    const past = this.search((i) => this.lines[i].start >= to);
    let chars = past > first ? this.sums[past] - this.sums[first] : 0;
    const unheard = to - Math.max(from, this.heard);
    if (unheard > 0) chars += unheard * (this.room?.rate ?? 0);
    return chars;
  }

  fits(from: number, to: number): boolean {
    return !this.known || this.weight(from, to) <= this.room!.chars;
  }

  // The latest end a window starting at from may have.
  longestFrom(from: number): number {
    if (this.fits(from, this.duration)) return this.duration;
    let lo = from;
    let hi = this.duration;
    for (let i = 0; i < halvings; i++) {
      const mid = (lo + hi) / 2;
      if (this.fits(from, mid)) lo = mid;
      else hi = mid;
    }
    // On a whole second, so the window can be said out loud.
    const whole = Math.floor(lo + 1e-6);
    return whole > from && this.fits(from, whole) ? whole : lo;
  }

  // The earliest start a window ending at to may have.
  earliestTo(to: number): number {
    if (this.fits(0, to)) return 0;
    let lo = 0;
    let hi = to;
    for (let i = 0; i < halvings; i++) {
      const mid = (lo + hi) / 2;
      if (this.fits(mid, to)) hi = mid;
      else lo = mid;
    }
    const whole = Math.ceil(hi - 1e-6);
    return whole < to && this.fits(whole, to) ? whole : hi;
  }

  // The longest window that fits wherever it is drawn: the shortest of the
  // longest windows from every start. It is what the clip settings are held
  // to, so the clips they ask for fit in a window drawn anywhere.
  //
  // The densest stretch decides it, and a window is densest when it starts
  // just before a line ends: the line is counted whole for a moment of it.
  // Past the transcript every second weighs the same, so one start there
  // says it all. A window cut short by the end of the
  // episode does not count, because the end is not the model's limit.
  anywhere(): number {
    if (!this.known || this.fits(0, this.duration)) return this.duration;
    let shortest = this.duration;
    const starts = [0, this.heard, ...this.lines.map((l) => Math.max(l.start, l.end - 0.001))];
    for (const start of starts) {
      if (start >= this.duration) continue;
      const end = this.longestFrom(start);
      if (end >= this.duration) continue;
      shortest = Math.min(shortest, end - start);
    }
    return shortest;
  }

  // The first index where test turns true, over lines in order, or the
  // number of lines when it never does.
  private search(test: (i: number) => boolean): number {
    let lo = 0;
    let hi = this.lines.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (test(mid)) hi = mid;
      else lo = mid + 1;
    }
    return lo;
  }
}

// The shortest a window may be: room for the clips asked for, one after
// another at their shortest, and never longer than the episode. The engine
// refuses less, see Holds in engine/room.go.
export function leastWindow(count: number, shortest: number, duration: number, floor = 10): number {
  return Math.min(Math.max(count * shortest, floor), duration);
}

// The most clips a search may ask for: as many as fit at their shortest in
// the longest window that can be drawn anywhere, and no more than cap.
export function mostClips(reach: number, shortest: number, cap: number): number {
  return Math.max(1, Math.min(cap, Math.floor((reach + 0.01) / Math.max(shortest, 1))));
}

// The longest the shortest clip may be: the clips asked for, at that length,
// still fit in that window. Never more than cap, and never less than floor,
// the least any clip may be.
export function longestShortest(reach: number, count: number, floor: number, cap: number): number {
  return Math.max(floor, Math.min(cap, Math.floor((reach + 0.01) / Math.max(count, 1))));
}

// A window put back inside what it may be: no longer than the model reads
// from where it starts, no shorter than its clips need. It keeps its start
// when it can, and gives up the end first, because the start is where the
// window was put.
export function fitWindow(
  win: { from: number; to: number },
  reach: Reach,
  least: number,
): { from: number; to: number } {
  const duration = reach.duration;
  let from = Math.min(Math.max(win.from, 0), duration);
  let to = Math.min(Math.max(win.to, from), duration);
  to = Math.min(to, reach.longestFrom(from));
  if (to - from < least) {
    to = Math.min(duration, from + least);
    from = Math.max(0, to - least);
  }
  return { from, to };
}
